### PR TITLE
IncomingPhoneNumbers

### PR DESCRIPTION
**Pull request by chelpu**

### CODE REVIEW

File: examples/modify_conference/main.go [hunk 1]
```diff
@@ -56,7 +60,7 @@ func main() {
 	bufio.NewReader(os.Stdin).ReadString('\n')
 	resp, err := act.Conferences.List(conference.ListFilter{
 		Status:      "in-progress",
-		DateCreated: &(common.JSONTime{time.Date(2014, time.July, 11, 3, 45, 01, 0, &time.Location{})}),
+		DateCreated: &(common.JSONTime{time.Date(2014, time.July, 12, 3, 45, 01, 0, &time.Location{})}), //Set to today's date
```
Comment (natebrennand) [inline examples/modify_conference/main.go]: Why don't you just use `time.Now()` here?
http://golang.org/pkg/time/#Now

File: numbers/numbers.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+		vals.Set("ApplicationSid", p.VoiceApplicationSid)
+	}
+	if p.SmsURL != "" {
+		vals.Set("SmsUrl", p.SmsURL)
+	}
+	if p.SmsMethod != "" {
+		vals.Set("SmsMethod", p.SmsMethod)
+	}
+	if p.SmsFallbackURL != "" {
+		vals.Set("SmsFallbackURL", p.SmsFallbackURL)
```
Comment (natebrennand) [inline numbers/numbers.go]: URL -> Url

File: numbers/numbers.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+		vals.Set("ApiVersion", p.APIVersion)
+	}
+	if p.VoiceURL != "" {
+		vals.Set("VoiceUrl", p.VoiceURL)
+	}
+	if p.VoiceMethod != "" {
+		vals.Set("VoiceMethod", p.VoiceMethod)
+	}
+	if p.VoiceFallbackURL != "" {
+		vals.Set("VoiceFallbackURL", p.VoiceFallbackURL)
```
Comment (natebrennand) [inline numbers/numbers.go]: URL -> Url

File: numbers/response.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	AccountSid           string          `json:"account_sid"`
+}
+
+type NumberList struct {
+	common.ListResponseCore
+	IncomingPhoneNumbers *[]Number `json:"incoming_phone_numbers"`
+}
+
+func validateNumberSid(sid string) bool {
+	match, _ := regexp.MatchString(`^PN[0-9a-z]{32}$`, sid)
```
Comment (natebrennand) [inline numbers/response.go]: can you do this in the new style?
You should be able to drop this in as a replacement:

``` go
var validateNumberSid = regexp.MustCompile(`^PN[0-9a-z]{32}$`).MatchString
```

File: voice/calls.go [hunk 5]
```diff
@@ -168,12 +155,8 @@ func (f ListFilter) GetQueryString() string {
...
 }
 
-func (act Account) getList(destURL string, f ListFilter, resp *CallList) error {
-	return common.SendGetRequest(destURL+f.GetQueryString(), act, resp)
-}
-
 func (act Account) List(f ListFilter) (CallList, error) {
 	var callList CallList
-	err := act.getList(fmt.Sprintf(listURL, act.AccountSid), f, &callList)
+	err := common.SendGetRequest(fmt.Sprintf(listURL, act.AccountSid)+f.GetQueryString(), act, &callList)
```
Comment (natebrennand) [inline voice/calls.go]: GetQueryString -> getQueryString ?
there's no reason to make this a public method

File: numbers/numbers.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+
+	err := common.SendGetRequest(fmt.Sprintf(getURL, act.AccountSid, numberSid), act, &p)
+	return p, err
+}
+
+// Updates properties of a phone number via a post
+func (act Account) Post(numberSid string, update Post) (Number, error) {
+	var p Number
+	if !validateNumberSid(numberSid) {
+		return p, errors.New("Invalid sid")
```
Comment (natebrennand) [inline numbers/numbers.go:179]: should we go ahead and move this logic to Post's Validate() method?

File: numbers/numbers.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+	}
+	return common.SendDeleteRequest(fmt.Sprintf(getURL, act.AccountSid, numberSid), act)
+}
+
+type ListFilter struct {
+	PhoneNumber  string
+	FriendlyName string
+}
+
+func (f ListFilter) GetQueryString() string {
```
Comment (natebrennand) [inline numbers/numbers.go]: this should also be a private method I think, no reason to expose it

File: numbers/numbers.go [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+func (act Account) Post(numberSid string, update Post) (Number, error) {
+	var p Number
+	if !validateNumberSid(numberSid) {
+		return p, errors.New("Invalid sid")
+	}
+	err := common.SendPostRequest(fmt.Sprintf(getURL, act.AccountSid, numberSid), update, act, &p)
+	return p, err
+}
+
+func (act Account) PostNumber(n NumberSelector) (Number, error) {
```
Comment (natebrennand) [inline numbers/numbers.go]: Could we be more descriptive than `PostNumber`, is this endpoint for creating a number?

File: numbers/numbers.go (hunk 6 above)
Comment (chelpu) [inline numbers/numbers.go:179]: Wouldn't I need to pass in the PNSid to validate in order to do that?

Comment (natebrennand) [inline numbers/numbers.go:179]: whoops, yes, nvm

File: conference/fixtures_test.go [hunk 9]
```diff
@@ -90,6 +104,18 @@ var testParticipantFixtureString = `{
...
 
+var testParticipantFixture = Participant{
+	AccountSid:          "AC5116d5d4df9f61ceae2f0732e1ea9f1b",
+	DateCreated:         common.JSONTime{time.Date(2010, time.August, 16, 3, 45, 01, 0, &time.Location{})},
+	DateUpdated:         common.JSONTime{time.Date(2010, time.August, 16, 3, 45, 03, 0, &time.Location{})},
+	ConferenceSid:       "CFbbe46ff1274e283f7e3ac1df0072ab39",
+	CallSid:             "CA386025c9bf5d6052a1d1ea42b4d16662",
+	EndConferenceOnExit: true,
+	Muted:               false,
+	StartConferenceOnEnter: true,
```
Comment (natebrennand) [inline conference/fixtures_test.go:115]: you still need to run `go fmt ./...`. add a pre-commit hook?

File: conference/participant.go [hunk 10]
```diff
@@ -49,8 +49,8 @@ func (act Account) Participant(confSid string, callSid string) (Participant, err
 	} else if !validateCallSid(callSid) {
 		return p, errors.New("Invalid call sid")
 	}
-
-	err := common.SendGetRequest(fmt.Sprintf(getURL, act.AccountSid, confSid, callSid), act, &p)
+	err := common.SendGetRequest(fmt.Sprintf(participantURL, act.AccountSid, confSid, callSid), act, &p)
+	fmt.Println("p: ", p.CallSid)
```
Comment (natebrennand) [inline conference/participant.go]: make sure to delete this line too

File: numbers/numbers.go [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+type Core struct {
+	FriendlyName         string
+	APIVersion           string
+	VoiceURL             string
+	VoiceMethod          string
+	VoiceFallbackURL     string
+	VoiceFallbackMethod  string
+	StatusCallback       string
+	StatusCallbackMethod string
+	VoiceCallerIdLookup  *bool
```
Comment (natebrennand) [inline numbers/numbers.go]: can this be a plain `bool`?

File: voice/calls.go (hunk 5 above)
Comment (natebrennand) [inline voice/calls.go]: bump for private method

File: conference/fixtures_test.go (hunk 9 above)
Comment (natebrennand) [inline conference/fixtures_test.go:115]: i don't understand `go fmt`..., guess it doesn't apply when the attribute names are 22 characters long

**Pull request merged**